### PR TITLE
feat: Add support for webhook servers

### PR DESCRIPTION
**Pull request by Techassi**
Fixes https://github.com/stackabletech/issues/issues/508, supersedes #695, part of https://github.com/stackabletech/issues/issues/361

This adds support to create webhook servers based on `axum`. It additionally provides a ready-to-use conversion webhook server.

---

This PR **does not** include code for TLS cert auto-generation and as such doesn't support it (marked with `todo!()`). The code for this feature will be handled by a follow-up PR. Furthermore, the code for TLS cert chains will also move into the new crate.

### CODE REVIEW

**Comment by Techassi** [comment]
The crates are now all up-to-date. This PR is now reviewable.

**Review by NickLarsenNZ — CHANGES_REQUESTED** [review]
I generally approve, but left enough comments that I assume changes will be made.

File: stackable-webhook/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+/// A generic webhook handler receiving a request and state and sending back
+/// a response.
+///
+/// This trait is usually not implemented by external callers and this library
+/// provides various ready-to-use implementations for it. One such an
+/// implementation is part of the [`ConversionWebhookServer`].
+pub trait StatefulWebhookHandler<Req, Res, S> {
+    fn call(self, req: Req, state: S) -> Res;
+}
```
[the remark this answers is not in the document]
Comment (Techassi) [reply stackable-webhook/src/lib.rs:60]: Yes, we could make it private. For now, there is no need to export it.

File: stackable-webhook/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+/// This trait is usually not implemented by external callers and this library
+/// provides various ready-to-use implementations for it. One such an
+pub trait WebhookHandler<Req, Res> {
+    fn call(self, req: Req) -> Res;
+}
+
+/// A generic webhook handler receiving a request and state and sending back
+/// a response.
+///
+/// This trait is usually not implemented by external callers and this library
```
[the remark this answers is not in the document]
Comment (Techassi) [reply stackable-webhook/src/lib.rs:55]: Same as above.

File: stackable-webhook/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+                let redirector = Redirector::new(
+                    self.options.socket_addr.ip(),
+                    self.options.socket_addr.port(),
+                    http_port,
+                );
+
+                info!(http_port, "spawning redirector in separate task");
+                tokio::spawn(redirector.run());
+            }
```
[the remark this answers is not in the document]
Comment (Techassi) [reply stackable-webhook/src/lib.rs:145]: I would be happy to remove it entirely, but let's see if other reviewers have different opinions.

File: stackable-webhook/src/options.rs [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+    /// Enables HTTP to HTTPS auto-redirect on `http_port`. The webhook
+    /// server will listen on both HTTP and HTTPS.
+    pub fn enable_redirect(mut self, http_port: u16) -> Self {
+        self.redirect = Some(RedirectOption::Enabled(http_port));
+        self
+    }
+
+    /// Sets the socket address the webhook server uses to bind for HTTPS.
+    pub fn socket_addr(mut self, socket_ip: impl Into<IpAddr>, socket_port: u16) -> Self {
```
[the remark this answers is not in the document]
Comment (Techassi) [reply stackable-webhook/src/options.rs:126]: Sure!

File: stackable-webhook/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        debug!("create core couter and merge provided router");
+        let mut router = Router::new();
+        router = router.merge(self.router);
+
+        // Create server for TLS termination
+        debug!("create TLS server");
+        let tls_server = TlsServer::new(self.options.socket_addr, router, self.options.tls)
+            .context(CreateTlsServerSnafu)?;
+
+        info!("running TLS server");
```
[the remark this answers is not in the document]
Comment (soenkeliebau) [reply stackable-webhook/src/lib.rs:161]: I think its a fairly common frame of mind regarding loglevels, yeah.

File: stackable-webhook/src/servers/conversion.rs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    #[instrument(name = "create_conversion_webhhok_server", skip(handler))]
+    pub fn new<H>(handler: H, options: Options) -> Self
+    where
+        H: WebhookHandler<ConversionReview, ConversionReview> + Clone + Send + Sync + 'static,
+    {
+        debug!("create new conversion webhook server");
+
+        let handler_fn = |Json(review): Json<ConversionReview>| async {
+            let review = handler.call(review);
+            Json(review)
```
[the remark this answers is not in the document]
Comment (Techassi) [reply stackable-webhook/src/servers/conversion.rs:68]: Nice idea, gonna add it as a middleware to the core router.

File: stackable-webhook/src/tls/certs.rs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+pub struct CertificateChain {
+    chain: Vec<CertificateDer<'static>>,
+    private_key: PrivateKeyDer<'static>,
+}
+
+impl CertificateChain {
+    pub fn from_files(
+        cert_path: impl AsRef<Path>,
+        pk_path: impl AsRef<Path>,
+        pk_encoding: PrivateKeyEncoding,
```
[the remark this answers is not in the document]
Comment (Techassi) [reply stackable-webhook/src/tls/certs.rs:35]: Will update it accordingly.

File: stackable-webhook/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Techassi) [reply stackable-webhook/src/lib.rs]: Changed in d0ea673

File: stackable-webhook/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Techassi) [reply stackable-webhook/src/lib.rs]: Changed in d0ea673

File: stackable-webhook/src/tls/certs.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Techassi) [reply stackable-webhook/src/tls/certs.rs:35]: Changed in 2e53410

File: stackable-webhook/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Techassi) [reply stackable-webhook/src/lib.rs:162]: Alright, I will change `info!()` to `debug!()`.

File: stackable-webhook/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+/// A result type alias with the library-level [`Error`] type as teh default
+/// error type.
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// A generic webhook handler receiving a request and sending back a response.
+///
+/// This trait is not intended to be implemented by external crates and this
+/// library provides various ready-to-use implementations for it. One such an
+/// /// implementation is part of the [`ConversionWebhookServer`].
```
Comment (NickLarsenNZ) [inline stackable-webhook/src/lib.rs:48]: Duplicate doc-comment thing

```suggestion
/// implementation is part of the [`ConversionWebhookServer`].
```

File: stackable-webhook/src/servers/conversion.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (NickLarsenNZ) [reply stackable-webhook/src/servers/conversion.rs:68]: I assume in a separate PR (not necessary to get this feature in).

[the remark this answers is not in the document]
Comment (Techassi) [reply stackable-webhook/src/servers/conversion.rs:68]: I added basic high-level request tracing support in e1f538b. I will add full OpenTelemetry compatible tracing in a follow-up PR.

File: stackable-webhook/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        debug!("create tracing service (layer)");
+        let layer = TraceLayer::new_for_http()
+            .make_span_with(|_: &Request<Body>| debug_span!("webhook_request"))
+            .on_body_chunk(())
+            .on_eos(());
+
+        let service = ServiceBuilder::new().layer(layer);
+
+        // Create the root router and merge the provided router into it.
+        debug!("create core couter and merge provided router");
```
Comment (NickLarsenNZ) [inline stackable-webhook/src/lib.rs:167]: nit: typo

```suggestion
        debug!("create core counter and merge provided router");
```

**Review by NickLarsenNZ — DISMISSED** [review]
LGTM

File: stackable-webhook/src/lib.rs (hunk 9 above)
Comment (Techassi) [reply stackable-webhook/src/lib.rs:145]: Actually, `router` :)

Comment (Techassi) [reply stackable-webhook/src/lib.rs]: Fixed in e9ffe0e

File: stackable-webhook/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Techassi) [reply stackable-webhook/src/lib.rs]: Removed in dbc9dd3

File: stackable-webhook/src/redirect.rs
[the remark this answers is not in the document]
Comment (Techassi) [reply stackable-webhook/src/redirect.rs:1]: Removed in dbc9dd3

File: stackable-webhook/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+                    self.options.socket_addr.ip(),
+                    self.options.socket_addr.port(),
+                    http_port,
+                );
+
+                info!(http_port, "spawning redirector in separate task");
+                tokio::spawn(redirector.run());
+            }
+            RedirectOption::Disabled => {
+                warn!("webhook runs without automatic HTTP to HTTPS redirect which is not recommended");
```
[the remark this answers is not in the document]
Comment (Techassi) [reply stackable-webhook/src/lib.rs]: Removed in dbc9dd3

File: stackable-webhook/src/options.rs [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+    tls::certs::PrivateKeyEncoding,
+};
+
+/// Specifies available webhook server options.
+///
+/// The [`Default`] implemention for this struct contains the following
+/// values:
+///
+/// - Redirect from HTTP to HTTPS is enabled, HTTP listens on port 8080
+/// - The socket binds to 127.0.0.1 on port 8443 (HTTPS)
```
[the remark this answers is not in the document]
Comment (Techassi) [reply stackable-webhook/src/options.rs]: Removed in dbc9dd3

**Review by NickLarsenNZ — APPROVED** [review]
Lessgooo

**Pull request merged**